modules/OFConnectionManager2/module/src/ofconnectionmanager.c

```diff
@@ -312,8 +328,164 @@ controller_t_init(controller_t *controller,
+    if (X509_verify_cert(verify_ctx) != 1) {
+        AIM_LOG_ERROR("Failed to verify certificate: %s",
+                      X509_verify_cert_error_string(verify_ctx->error));
+        goto error;
+    }
+
+    is_verified = true;
+
+ error:
+    if (cert) {
+        X509_free(cert);
+    }
+    if (fp_cert) {
+        fclose(fp_cert);
+    }
+    if (verify_ctx) {
+        X509_STORE_CTX_free(verify_ctx);
+    }
+    
+    return is_verified;
+}
+
+
+static SSL_CTX *
+ssl_ctx_alloc(void)
+{
+    SSL_CTX *ctx;
+    char buf[IND_SSL_ERR_LEN];
+
+    ctx = SSL_CTX_new(TLSv1_2_client_method());
+    if (ctx == NULL) {
+        ERR_error_string(ERR_get_error(), buf);
+        AIM_LOG_ERROR("Failed to allocate SSL_CTX: %s", buf);
+        goto error;
+    }
+    if (SSL_CTX_set_cipher_list(ctx, tls_cfg.cipher_list) != 1) {
+        ERR_error_string(ERR_get_error(), buf);
+        AIM_LOG_ERROR("Failed to set cipher list: %s", buf);
+        goto error;
+    }
+    if (SSL_CTX_load_verify_locations(ctx, tls_cfg.ca_cert, NULL) != 1) {
+        ERR_error_string(ERR_get_error(), buf);
+        AIM_LOG_ERROR("Failed to set CA certificate file: %s", buf);
+        goto error;
+    }
+    if (SSL_CTX_use_certificate_file(ctx, tls_cfg.switch_cert,
+                                     SSL_FILETYPE_PEM) != 1) {
+        ERR_error_string(ERR_get_error(), buf);
+        AIM_LOG_ERROR("Failed to set certificate file: %s", buf);
+        goto error;
+    }
+    if (SSL_CTX_use_PrivateKey_file(ctx, tls_cfg.switch_priv_key,
+                                    SSL_FILETYPE_PEM) != 1) {
+        ERR_error_string(ERR_get_error(), buf);
+        AIM_LOG_ERROR("Failed to set private key file: %s", buf);
+        goto error;
+    }
+    if (SSL_CTX_check_private_key(ctx) != 1) {
+        ERR_error_string(ERR_get_error(), buf);
+        AIM_LOG_ERROR("Private key does not match public certificate: %s", 
+                      buf);
+        goto error;
+    }
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
+
+    /* verify certificate chain */
+    if (!verify_cert(ctx, tls_cfg.switch_cert)) {
+        AIM_LOG_ERROR("Failed to verify certificate file");
+        goto error;
+    }
+
+    return ctx;
+
+ error:
+    if (ctx) {
+        SSL_CTX_free(ctx);
+        ctx = NULL;
+    }
+    return NULL;
+}
+
+static void
+ssl_ctx_free(SSL_CTX *ctx)
+{
+    SSL_CTX_free(ctx);
+}
+
+
+/* public API for TLS configuration */
+indigo_error_t
+indigo_cxn_config_tls(char *cipher_list,
+                      char *ca_cert,
+                      char *switch_cert,
+                      char *switch_priv_key)
+{
+    SSL_CTX *ctx;
+
+    strncpy(tls_cfg.cipher_list, cipher_list, sizeof(tls_cfg.cipher_list));
+    strncpy(tls_cfg.ca_cert, ca_cert, sizeof(tls_cfg.ca_cert));
+    strncpy(tls_cfg.switch_cert, switch_cert, sizeof(tls_cfg.switch_cert));
+    strncpy(tls_cfg.switch_priv_key, switch_priv_key,
+            sizeof(tls_cfg.switch_priv_key));
+
+    ctx = ssl_ctx_alloc();
```

Comment:
Suggest refactoring ssl_ctx_alloc to take a tls_cfg_t so that the parameter checking can be done before the global configuration is overwritten.